utils/resb/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! `resb` is a utility crate of the [`ICU4X`] project for working with ICU
+//! resource bundle files.
+//!
+//! It comprises modules for reading and optionally writing [`binary`] `.res`
+//! files as well as optionally for reading [`text`] bundles.
+//!
+//! [`ICU4X`]: ../icu/index.html
+
+// https://github.com/unicode-org/icu4x/blob/main/docs/process/boilerplate.md#library-annotations
+#![cfg_attr(not(any(test, feature = "std")), no_std)]
+#![cfg_attr(
+    not(test),
+    deny(
+        clippy::indexing_slicing,
+        clippy::unwrap_used,
+        clippy::expect_used,
+        clippy::panic,
+        clippy::exhaustive_structs,
+        clippy::exhaustive_enums,
+        missing_debug_implementations,
+    )
+)]
+#![warn(missing_docs)]
+
+use core::fmt;
+
+extern crate alloc;
+use alloc::borrow::Cow;
+
+/// The `binary` module provides a `serde` deserializer for the binary resource
+/// bundle format as well as a means of writing resource bundles to the binary
+/// file format.
+pub mod binary;
+
+/// The `text` module provides a reader for the text-based resource bundle
+/// format.
+///
+/// WARNING: This reader is intended for use in development-time tools and is
+/// not written with runtime efficiency in mind.
+#[cfg(feature = "tools")]
+pub mod text;
+
+/// The `bundle` module provides data structures for working directly with the
+/// contents of a resource bundle.
+///
+/// WARNING: This module is not suitable for use at runtime due to its reliance
+/// on `std` and `alloc` and therefore not intended for general deserialization
+/// of resource bundles. Rather, it is intended to be used in development-time
+/// tools for working with bundles.
+#[cfg(feature = "std")]
+pub mod bundle;
+
+const MASK_28_BIT: u32 = 0x0fff_ffff;
+
+/// The `Error` struct provides basic error handling for serialization and
+/// deserialization of the ICU resource bundle formats.
+#[derive(Debug)]
+pub struct Error {
```

Comment:
Nit: Our Error types in ICU4X should be `Copy`, because it makes things smaller and faster and easier to optimize Result return values. I see you are doing string formatting like

`"Unexpected error. Expected resource of type {resource} to have been processed already."`

and

`"String of length {length} is too large."`

The ICU4X style is instead to make `Error` an `enum` and then have a bunch of variants that can carry additional data, and to `#[derive(Display)]` with the `displaydoc` crate to allow formatting of the nicer messages.

Reply:
I'll move `Error` to an `enum`, or maybe an `enum` of `enum`s to represent errors from different parts of the code, if that makes sense. For a lot of the errors, though, I think some additional debugging context would be useful. Would `&str` make sense for that?

Reply:
`&'static str` is fine, as are all types of primitives. If you need an error that is like "string of {length} is too large" you should just make it an enum variant with a usize field.

Reply:
Addressed.